### DISCUSSION TITLE
Override reported free space

### DISCUSSION THREAD

**Discussion by HiFiPhile** [drakkan/sftpgo#825] [Q&A]
Hi,

Long time no see :), is everything ok ?

I've a small issue in my new configuration:

### Server
- Debian Busseye
- SFTPGo 2.2.2-a792150-2022-02-06T14:12:25Z +metrics +azblob +gcs +s3 +bolt +mysql +pgsql +sqlite +portable

SFTPGo's working directory is set to `/var/run` where no file will be stored, multiple ZFS pools shared as external folders.

```
$ df -h
192.168.10.1:/P4510                    5.6T  2.3T  3.4T  41% /media/zfs/p4510
192.168.10.1:/NAS/repo                 5.0T   63G  5.0T   2% /media/zfs/repo
192.168.10.1:/S860                     790G   92G  698G  12% /media/zfs/ssd
192.168.10.1:/home/log                 889G  447G  443G  51% /media/log
192.168.10.1:/NAS/media                 16T   11T  5.0T  68% /media/zfs/media
192.168.10.1:/NAS/data                 6.5T  1.6T  5.0T  25% /media/zfs/data
192.168.10.1:/NAS/music                5.5T  529G  5.0T  10% /media/zfs/music
192.168.10.1:/S860/BT                  7.1T  6.4T  698G  91% /media/zfs/ssd/BT
192.168.10.1:/mnt/ramdisk              4.0G   22M  4.0G   1% /media/ramdisk
192.168.10.1:/P4510/BT                 3.9T  562G  3.4T  15% /media/zfs/p4510/BT
```

### Client
- archlinux
- SSHFS version 3.7.2
FUSE library version 3.10.5
using FUSE kernel interface version 7.31
fusermount3 version: 3.10.5
- KDE dolphin 21.12.3

```
$ df -h
ftp@xxxxxx.com:  1.6G  1.3M  1.6G   1% /media/sftp
```

### Issue
When I copy big files to server with Dolphin, it fails with error "Not enough free space". 

I think it's because Dolphin check free space before copy, since free space shown from client is wrong (it shows free space of SFTPGo's working directory). 

If I copy with `cp` then this check is bypassed.

So is there a simple way to override the reported free space ? For example with a Windows rclone client it report `1.0 PB` which is overkill.

**Comment by drakkan** [comment]
Hi,

I'm fine hope the same for you.

I'm not sure if there is a bug somewhere in SFTPGo, but the client must specify a path when it requests the available disk space and SFTPGo should check the space for the requested path.

Unfortunately it seems there are no logs for this request, can you please debug the issue and report back the results? The disk space request is handled [here](https://github.com/drakkan/sftpgo/blob/main/sftpd/handler.go#L272). Thanks!

**Comment by HiFiPhile** [comment]
I did some tests with `df`, I think SFTPGo handles free space correctly:
`df -h .` give me the free space of SFTPGo's working directory `/var/run`
`df -h ./xxx` give me free space of external folders.

I think it's a KDE Dolphin's limitation, it tests only the free space of root mount point, rather the destination folder.

**Reply by drakkan** [reply]
Thanks for your tests, so the issue seems unrelated to SFTPGo. Let me know if you find something that can be improved on the SFTPGo side

**Reply by HiFiPhile** [reply]
I found a workaround of this issue. 

As disk space is tested only if no quota is set [L290](https://github.com/drakkan/sftpgo/blob/1e0b3a2a8c1e48852ccb37ee035946be0a36bb69/sftpd/handler.go#L290), I set working directory's quota to 100TB and now KDE is happy ;)

**Reply by drakkan** [reply]
This looks like a bug, we should check the available disk space anyway and returns the minimum ... I'll check the next weekend, I'm too busy now

**Reply by drakkan** [reply]
This [commit](https://github.com/drakkan/sftpgo/commit/d2f4edcdb6507d351da2deaced94932311aa3786) will break your workaround, sorry :smile:

**Reply by HiFiPhile** [reply]
Yeah it's better to have it fixed.

Do you think it's useful to add an option to disable free space detection ? I can report the bug to KDE but maybe it needs years to fix. I don't know if there are other softwares who also only check the free space of root directory.

**Reply by drakkan** [reply]
Please try to report the bug to kde and post the link here. Based on their answer I will try to figure out what the best solution is, thanks

**Reply by HiFiPhile** [reply]
I've posted the bug report to KDE: https://bugs.kde.org/show_bug.cgi?id=453856

**Reply by drakkan** [reply]
> I've posted the bug report to KDE: https://bugs.kde.org/show_bug.cgi?id=453856

uhmm ... still no answer ... they don't seem very responsive

**Reply by HiFiPhile** [reply]
Yes, there are many bugs unanswered...

**Reply by HiFiPhile** [reply]
Same issue exist on Windows using https://github.com/winfsp/sshfs-win.
![image](https://user-images.githubusercontent.com/4375114/178974613-19cc2ebd-4144-4cf2-967d-82e9da77df33.png)

**Reply by drakkan** [reply]
> Same issue exist on Windows using https://github.com/winfsp/sshfs-win. ![image](https://user-images.githubusercontent.com/4375114/178974613-19cc2ebd-4144-4cf2-967d-82e9da77df33.png)



> Same issue exist on Windows using https://github.com/winfsp/sshfs-win. ![image](https://user-images.githubusercontent.com/4375114/178974613-19cc2ebd-4144-4cf2-967d-82e9da77df33.png)

This looks like a bug on their side. Both KDE and sshfs-win have a lot of sponsorship and so resources to fix their bugs. I'm not going to add an hack/workaround that I have to maintain forever, sorry.
